File: android/src/main/kotlin/com/morbit/tencent_map/TencentMapApi.kt
package com.morbit.tencent_map

import com.tencent.map.geolocation.TencentLocationManager
import com.tencent.tencentmap.mapsdk.maps.CameraUpdateFactory
import com.tencent.tencentmap.mapsdk.maps.TencentMap.MAP_TYPE_DARK
import com.tencent.tencentmap.mapsdk.maps.TencentMap.MAP_TYPE_NORMAL
import com.tencent.tencentmap.mapsdk.maps.TencentMap.MAP_TYPE_SATELLITE
import com.tencent.tencentmap.mapsdk.maps.TencentMapInitializer
import com.tencent.tencentmap.mapsdk.maps.model.LatLngBounds

class TencentMapApi(private val tencentMap: TencentMap) {
  private val mapView = tencentMap.view

  fun updateMapConfig(config: MapConfig) {
    config.mapType?.let {
      mapView.map.mapType = when (it) {
        MapType.NORMAL -> MAP_TYPE_NORMAL
        MapType.SATELLITE -> MAP_TYPE_SATELLITE
        MapType.DARK -> MAP_TYPE_DARK
      }
    }
    config.mapStyle?.let {
      mapView.map.mapStyle = it.toInt()
    }
    config.logoScale?.let { mapView.map.uiSettings.setLogoScale(it.toFloat()) }
    config.logoPosition?.let {
      mapView.map.uiSettings.setLogoPosition(
        it.anchor.toAnchor(),
        intArrayOf(it.offset.y.toInt(), it.offset.x.toInt())
      )
    }
    config.scalePosition?.let {
      mapView.map.uiSettings.setScaleViewPositionWithMargin(
        it.anchor.toAnchor(),
        it.offset.y.toInt(),
        it.offset.y.toInt(),
        it.offset.x.toInt(),
        it.offset.x.toInt()
      )
    }
    config.compassOffset?.let {
      mapView.map.uiSettings.setCompassExtraPadding(
        it.x.toInt(),
        it.y.toInt()
      )
    }
    config.compassEnabled?.let { mapView.map.uiSettings.isCompassEnabled = it }
    config.scaleEnabled?.let { mapView.map.uiSettings.isScaleViewEnabled = it }
    config.scaleFadeEnabled?.let { mapView.map.uiSettings.setScaleViewFadeEnable(it) }
    config.skewGesturesEnabled?.let { mapView.map.uiSettings.isTiltGesturesEnabled = it }
    config.scrollGesturesEnabled?.let { mapView.map.uiSettings.isScrollGesturesEnabled = it }
    config.rotateGesturesEnabled?.let { mapView.map.uiSettings.isRotateGesturesEnabled = it }
    config.zoomGesturesEnabled?.let { mapView.map.uiSettings.isZoomGesturesEnabled = it }
    config.trafficEnabled?.let { mapView.map.isTrafficEnabled = it }
    config.indoorViewEnabled?.let { mapView.map.setIndoorEnabled(it) }
    config.indoorPickerEnabled?.let { mapView.map.uiSettings.isIndoorLevelPickerEnabled = it }
    config.buildingsEnabled?.let { mapView.map.showBuilding(it) }
    config.buildings3dEnabled?.let { mapView.map.setBuilding3dEffectEnable(it) }
    config.myLocationEnabled?.let { mapView.map.isMyLocationEnabled = it }
    config.userLocationType?.let {
      if (mapView.map.isMyLocationEnabled) {
        mapView.map.setMyLocationStyle(it.toMyLocationStyle())
      }
    }
  }

  fun moveCamera(position: CameraPosition, duration: Long) {
    val cameraPosition = position.toCameraPosition(mapView.map.cameraPosition)
    val cameraUpdate = CameraUpdateFactory.newCameraPosition(cameraPosition)
    if (duration > 0) {
      mapView.map.stopAnimation()
      mapView.map.animateCamera(cameraUpdate, duration, null)
    } else {
      mapView.map.moveCamera(cameraUpdate)
    }
  }

  fun moveCameraToRegion(region: Region, padding: EdgePadding, duration: Long) {
    val latLngBounds = region.toLatLngBounds()
    val cameraUpdate = CameraUpdateFactory.newLatLngBoundsRect(
      latLngBounds,
      padding.left.toInt(),
      padding.right.toInt(),
      padding.top.toInt(),
      padding.bottom.toInt(),
    )
    if (duration > 0) {
      mapView.map.stopAnimation()
      mapView.map.animateCamera(cameraUpdate, duration, null)
    } else {
      mapView.map.moveCamera(cameraUpdate)
    }
  }

  fun moveCameraToRegionWithPosition(positions: List<Position?>, padding: EdgePadding, duration: Long) {
    val latLngBounds = LatLngBounds.Builder().include(positions.filterNotNull().map { it.toPosition() }).build()
    val cameraUpdate = CameraUpdateFactory.newLatLngBoundsRect(
      latLngBounds,
      padding.left.toInt(),
      padding.right.toInt(),
      padding.top.toInt(),
      padding.bottom.toInt(),
    )
    if (duration > 0) {
      mapView.map.stopAnimation()
      mapView.map.animateCamera(cameraUpdate, duration, null)
    } else {
      mapView.map.moveCamera(cameraUpdate)
    }
  }

  fun setRestrictRegion(region: Region, mode: RestrictRegionMode) {
    mapView.map.setRestrictBounds(
      region.toLatLngBounds(),
      mode.toRestrictMode()
    )
  }

  fun addMarker(marker: Marker) {
    val tencentMarker = mapView.map.addMarker(marker.toMarkerOptions(tencentMap.binding))
    tencentMap.markers[marker.id] = tencentMarker
    tencentMap.tencentMapMarkerIdToDartMarkerId[tencentMarker.id] = marker.id
  }

  fun removeMarker(id: String) {
    val marker = tencentMap.markers[id]
    if (marker != null) {
      marker.remove()
      tencentMap.markers.remove(id)
      tencentMap.tencentMapMarkerIdToDartMarkerId.remove(marker.id)
    }
  }

  fun updateMarker(markerId: String, options: MarkerUpdateOptions) {
    if (options.position != null) {
      tencentMap.markers[markerId]?.position = options.position.toPosition()
    }
    if (options.alpha != null) {
      tencentMap.markers[markerId]?.alpha = options.alpha.toFloat()
    }
    if (options.rotation != null) {
      tencentMap.markers[markerId]?.rotation = options.rotation.toFloat()
    }
    if (options.zIndex != null) {
      tencentMap.markers[markerId]?.zIndex = options.zIndex.toInt()
    }
    if (options.draggable != null) {
      tencentMap.markers[markerId]?.isDraggable = options.draggable
    }
    if (options.icon != null) {
      options.icon.toBitmapDescriptor(tencentMap.binding)?.let { tencentMap.markers[markerId]?.setIcon(it) }
    }
    if (options.anchor != null) {
      tencentMap.markers[markerId]?.setAnchor(options.anchor.x.toFloat(), options.anchor.y.toFloat())
    }
  }

  fun getUserLocation(): Location {
    return mapView.map.myLocation.toLocation()
  }

  fun start() {
    mapView.onStart()
  }

  fun pause() {
    mapView.onPause()
  }

  fun resume() {
    mapView.onResume()
  }

  fun stop() {
    mapView.onStop()
  }

  fun destroy() {
    mapView.onDestroy()
  }
}

File: ios/Classes/TencentMapApi.swift
import QMapKit
import Flutter

class _TencentMapApi: NSObject {
  let mapView: QMapView
  var markers = [String: QPointAnnotation]()

  init(mapView: QMapView) {
    self.mapView = mapView
  }

  func updateMapConfig(config: MapConfig) {
    if let type = config.mapType {
      mapView.mapType = [
        MapType.dark: QMapType.dark,
        MapType.normal: QMapType.standard,
        MapType.satellite: QMapType.satellite,
      ][type] ?? QMapType.standard
    }
    if let index = config.mapStyle {
      mapView.setMapStyle(Int32(index))
    }
    if let scale = config.logoScale {
      mapView.setLogoScale(scale)
    }
    if let position = config.logoPosition {
      mapView.setLogoMargin(position.offset.offset, anchor: position.anchor.anchor)
    }
    if let position = config.scalePosition {
      mapView.setScaleOffset(CGPointMake(position.offset.x, -position.offset.y))
    }
    if let offset = config.compassOffset {
      mapView.setCompassOffset(offset.offset)
    }
    if let enabled = config.compassEnabled {
      mapView.showsCompass = enabled
    }
    if let enabled = config.scaleEnabled {
      mapView.showsScale = enabled
    }
    if let enabled = config.scaleFadeEnabled {
      mapView.setScaleFadeEnable(enabled)
    }
    if let enabled = config.rotateGesturesEnabled {
      mapView.isRotateEnabled = enabled
    }
    if let enabled = config.scrollGesturesEnabled {
      mapView.isScrollEnabled = enabled
    }
    if let enabled = config.zoomGesturesEnabled {
      mapView.isZoomEnabled = enabled
    }
    if let enabled = config.skewGesturesEnabled {
      mapView.isOverlookingEnabled = enabled
    }
    if let enabled = config.indoorViewEnabled {
      mapView.setIndoorEnabled(enabled)
    }
    if let enabled = config.indoorPickerEnabled {
      mapView.indoorPicker = enabled
    }
    if let enabled = config.trafficEnabled {
      mapView.showsTraffic = enabled
    }
    if let enabled = config.buildingsEnabled {
      mapView.showsBuildings = enabled
    }
    if let enabled = config.buildings3dEnabled {
      mapView.shows3DBuildings = enabled
    }
    if let enabled = config.myLocationEnabled {
      mapView.showsUserLocation = enabled
    }
  }

  func moveCamera(position: CameraPosition, duration: Int64) {
    let animated = duration > 0
    if let it = position.position?.coordinate { mapView.setCenter(it, animated: animated) }
    if let it = position.zoom { mapView.setZoomLevel(CGFloat(it), animated: animated) }
    if let it = position.skew { mapView.setOverlooking(CGFloat(it), animated: animated) }
    if let it = position.heading { mapView.setRotation(CGFloat(it), animated: animated) }
  }

  func moveCameraToRegion(region: Region, padding: EdgePadding, duration: Int64) {
    mapView.setRegion(region.region, edgePadding: padding.padding, animated: duration > 0)
  }

  func moveCameraToRegionWithPosition(positions: [Position?], padding: EdgePadding, duration: Int64) {
    let coordinates = positions.filter { position in position != nil }.map { position in position!.coordinate }
    let coordinatesPointer = UnsafeMutablePointer<CLLocationCoordinate2D>.allocate(capacity: coordinates.count)
    coordinatesPointer.initialize(from: coordinates, count: coordinates.count)
    let region = QBoundingCoordinateRegionWithCoordinates(coordinatesPointer, UInt(coordinates.count))
    mapView.setRegion(region, animated: duration > 0)
  }

  func setRestrictRegion(region: Region, mode: RestrictRegionMode) {
    mapView.setLimitMapRect(QMapRectForCoordinateRegion(region.region), mode: mode.restrictMode)
  }

  func removeRestrictRegion() {
    mapView.setLimitMapRect(QMapRect(), mode: QMapLimitRectFitMode.width)
  }

  func addMarker(marker: Marker) {
    let annotation = marker.annotation
    markers[marker.id] = annotation
    mapView.addAnnotation(annotation)
  }

  func removeMarker(id: String) {
    if let annotation = markers[id] {
      mapView.removeAnnotation(annotation)
      markers.removeValue(forKey: id)
    }
  }

  func updateMarker(markerId: String, options: MarkerUpdateOptions) {
    if var annotation = markers[markerId] {
      mapView.removeAnnotation(annotation)
      var marker = annotation.marker(markerId: markerId)
      marker = marker.update(options)
      annotation = marker.annotation
      mapView.addAnnotation(annotation)
    }
  }

  func getUserLocation() -> Location {
    return mapView.userLocation.toLocation
  }

  func start() { }

  func pause() { }

  func resume() { }

  func stop() { }

  func destroy() { }
}
